### DISCUSSION TITLE
자료구조 - 그래프 - 호제&진호

### DISCUSSION THREAD

**Discussion by zini9188** [Shortudy-10th/tech-for-developer#66] [질의 응답]
**❓️ 
아래 2가지 질문에 답해주세요.
그래프 IS A 트리. 참 거짓여부와 그 이유
트리 IS A 그래프. 참 거짓여부와 그 이유**

**❓️프로그램상에서 그래프를 구현한다면, 어떤 방법으로 구현할 수 있을까요?**

**❓️대표적인 그래프 탐색법 두 개에 대해 간략하게 말해주세요.**

**❓️완전 그래프와 연결 그래프의 차이점은 무엇일까요?**

**❓️최소신장트리에 대해 설명해주세요.**

**❓️다익스트라 알고리즘과 벨만포드 알고리즘에 대해 설명하고 두 알고리즘의 차이점을 설명해주세요**

**❓️인접 행렬과 인접 리스트의 장단점은 무엇인가요?**

**Comment by Jihn0118** [comment]
**❓️인접 행렬과 인접 리스트의 장단점은 무엇인가요?**

**💬 답변**

1. 인접행렬(Adjancency Matrix)
- 장점
   - 2차원 배열에 **모든 정점들의 간선 정보가 있기** 때문에, 두 정점을 연결하는 간선을 조회할 때 O(1) 시간복잡도로 가능
- 단점
   - 간선의 수와 무관하게 항상 n² 크기의 2차원 배열이 필요하므로 **메모리 공간이 낭비**된다.
   - 그래프의 모든 간선의 수를 알아내려면 인접 행렬 전체를 확인해야 하므로 **O(n²)** 의 시간이 소요된다.

2. 인접 리스트(Adjacency List)
- 장점
   - 존재하는 간선만 관리하므로 보다 **메모리 사용이 효율적** 이다.
- 단점
   - 두 노드를 연결하는 간선을 조회하거나 노드의 차수를 알기 위해서는 노드의 인접 리스트를 탐색해야 하므로 **노드의 차수 만큼의 시간이 필요** 하다. 

![images_falling_star3_post_33d21ba3-12b3-4985-a28f-83a255c2e6e7_다운로드 (4)](https://github.com/Shortudy-10th/tech-for-developer/assets/87293724/9c231313-11d3-4212-82ea-4080db196281)

**Comment by Jihn0118** [comment]
**❓️프로그램상에서 그래프를 구현한다면, 어떤 방법으로 구현할 수 있을까요?**

**💬 답변**
그래프를 구현하는 방법에는 **배열(Array)** 을 사용하는 방법과 **연결리스트(Linked List)** 를 사용하는 방법이 있다.

1. 배열을 이용한 인접 행렬
노드의 개수가 n이라면 n * n 형태의 **2차원 배열** 로 그래프의 연결 관계로 표현
인접 행렬에서 행과 열은 노드를 의미하며, 각각의 원소들은 노드 간의 간선을 나
타낸다.
adj[i][j]: 노드 i에서 노드 j로 가는 간선이 있으면 1, 없으면 0
![images_falling_star3_post_c20e690d-9674-49e7-b450-c3e15ed36268_22](https://github.com/Shortudy-10th/tech-for-developer/assets/87293724/40d8d396-31b2-4c2e-84f5-45a15351638c)

- 가중치가 없는 무방향 그래프 
연결되어 있는 경우 행렬에서 1의 값을 가지고, 연결되지 않은 경우 0의 값을 가진다. 
두 개의 노드에서 간선이 동시에 연결되어 있기 때문에 **인접 행렬이 대칭적 구조** 를 가진다.

- 가중치가 있는 유방향 그래프
행렬에서 **각 간선의 가중치 값이 저장** 된다. 
이 경우 가중치가 0인 것과 간선이 없는 것이 구별되어야 한다.

---------------------------

2. 연결리스트를 이용한 인접 리스트
그래프의 각 노드에 인접한 노드들을 **연결리스트(Linked List)** 로 표현하는 방법이다.
즉, 노드의 개수만큼 인접리스트가 존재하며, 각각의 인접리스트에는 인접한 노드 정보가 저장된다. 그래프는 각 인접리스트에 대한 헤드포인터를 배열로 갖는다.
       - 헤드포인터: 연결 리스트의 맨 처음 노드를 가리키는 포인터
![images_falling_star3_post_fe4c1a20-1b9a-4f64-ab72-db43276a050d_22](https://github.com/Shortudy-10th/tech-for-developer/assets/87293724/4fbd6533-25f8-46c9-9dcd-891ccfed11f3)

- 가중치가 없는 무방향 그래프
그림과 같이 가중치 표현 없이 인접한 노드 정보가 저장된다.

- 가중치가 있는 유방향 그래프
종점 [노드 번호 | 간선의 가중치] 정보가 저장된다.

**Comment by Jihn0118** [comment]
**❓️다익스트라 알고리즘과 벨만포드 알고리즘에 대해 설명하고 두 알고리즘의 차이점을 설명해주세요**

**💬 답변**

1. 다익스트라(Dijkstra) 알고리즘
그래프 상에서 특정 한 노드에서 다른 모든 노드까지의 최단 거리를 구하는 알고리즘
(가중 그래프에서 간선 가중치의 합이 최소가 되는 경로를 찾는 알고리즘 중 하나)
그리디 + DP가 합쳐진 형태 
현재 위치한 노드에서 최선의 경로를 반복적으로 찾으면서도 계산 해둔 경로를 활용해 중복된 하위 문제를 푼다. 

2. 벨만 포드(Bellman Ford) 알고리즘
가중 유방향 그래프 상에서 특정 한 노드로부터 다른 노드까지의 최단 경로를 구하는 알고리즘
벨만 포드가 다익스트라의 한계점을 보완하기 위해 나왔다. 
한계점은 간선이 음수일 때 최단 경로를 구할 수 없다는 것.
정확히는 음의 가중치 자체는 문제가 되진 않는데 사이클을 형성하면 문제가 된다.

둘의 가장 큰 차이점은 **간선의 가중치가 음일 때 최소 비용을 구할 수 있는 것** 은 **벨만 포드**, 구할 수 없는 것은 **다익스트라**이다.

다만 시간 복잡도가 늘어나기 때문에 **가중치가 모두 양수**일 경우 **다익스트라**를 사용하는 것이 좋다. 
시간 복잡도가 늘어나는 이유는 그리디하게 최소 비용 경로를 찾아가는 다익스트라와 달리, 벨만 포드는 모든 경우의 수를 고려하는 DP가 사용되었기 때문이다.

**Comment by zini9188** [comment]
❓️대표적인 그래프 탐색법 두 개에 대해 간략하게 말해주세요.

![1-3544f094bd8c3215f1d8f70e0ddef0b1](https://github.com/Shortudy-10th/tech-for-developer/assets/83208807/d6fdf52b-fba7-42ee-ac0a-a8fc92290819)

💬 답변
### bfs
- 너비 우선 탐색
- 현재 정점에 연결된 가까운 점들부터 탐색
- 큐를 이용한 구현

### dfs
- 깊이 우선 탐색
- 현재 정점에서 갈 수 있는 점들까지 들어가며 탐색
- 스택 또는 재귀 함수로 구현

**Comment by zini9188** [comment]
❓️
아래 2가지 질문에 답해주세요.
그래프 IS A 트리. 참 거짓여부와 그 이유
트리 IS A 그래프. 참 거짓여부와 그 이유

💬 답변

![image](https://github.com/Shortudy-10th/tech-for-developer/assets/83208807/d920ed74-f7f4-4234-92bd-3da76a6925e1)

그래프는 객체들 사이의 이진 관계를 모델링한 수학적인 구조이다. 그래프는 정점과, 간선으로 이루어져 있다.

트리는 그래프의 일종으로 특별한 속성을 가진다. 
1. 순환 구조가 없어야 한다.
2. 모든 정점들이 연결되어 있어야 한다. 즉 연결 그래프여야 한다.
3. 정점의 수가 n개 일 때 간선의 수는 n - 1개여야 한다.

트리 IS A 그래프는 참이다.
그래프 IS A 트리는 거짓이다.

**Comment by zini9188** [comment]
❓️최소신장트리에 대해 설명해주세요.

💬 답변

최소 신장 트리의 목적은 무방향 가중치 그래프 중에서 간선 가중치의 합이 최소인 트리를 찾는 것이다.
실생활에서 네트워크 구축, 전기 회로 연결 등의 최적화 문제를 해결할 때 사용한다.
대표적인 최소 신장 트리 알고리즘으로는 크루스칼 알고리즘과 프림 알고리즘이 있다.

### 크루스칼 알고리즘
- 간선을 가중치 순으로 정렬한 후, 순서대로 간선을 추가하면서 트리를 생성한다. 이때 순환을 생성하는 간선은 추가하지 않는다. 이를 위해 서로소 집합 자료구조를 사용할 수 있다.

### 프림 알고리즘
- 임의의 정점에서 시작하여, 방문하지 않은 정점 중 최소 가중치 간선으로 연결되는 정점을 선택하여 트리를 확장한다. 우선순위 큐를 사용하여 효율적인 처리가 가능하다.

**Comment by zini9188** [comment]
❓️완전 그래프와 연결 그래프의 차이점은 무엇일까요?

### 완전 그래프
그래프 내의 모든 정점이 서로 직접 연결된 경우를 말한다. 즉 간선의 개수가 최대인 그래프이다. 모든 정점 쌍 사이에 간선이 존재하여야 하며, 간선 개수는 n(n-1)/2개이다. 

### 연결 그래프
그래프 내의 모든 정점이 서로 간접적으로라도 접근 가능한 경우를 말한다. 즉 그래프 내에서 임의의 두 정점 사이에 경로가 존재하면 연결 그래프이다. 간선 개수는 적어도 n-1개이고, 최대 n(n-1)/2개이다.

종류 | 간선의 개수 | 특징
-- |  -- | --
완전 그래프 | n(n-1)/2 (최대) | 모든 정점 쌍 사이에 간선이 존재함
연결 그래프 | n-1 (최소) ~ n(n-1)/2 (최대) | 임의의 두 정점 사이에 경로 존재
